File: solutions/Java/WordBreakDpTrie.java
import java.util.*;

class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean is_end = false;
}

/**
 * Determine if string can be segmented using words from dictionary (Trie approach).
 *
 * Time Complexity: O(n*m)
 * Space Complexity: O(n + T)
 */
public class WordBreakDpTrie {
    public static boolean wordBreakDpTrie(String s, List<String> wordDict) {
        TrieNode root = new TrieNode();
        for (String word : wordDict) {
            TrieNode node = root;
            for (char ch : word.toCharArray()) {
                node.children.putIfAbsent(ch, new TrieNode());
                node = node.children.get(ch);
            }
            node.is_end = true;
        }

        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;

        for (int i = 1; i <= s.length(); i++) {
            if (!dp[i]) {
                TrieNode node = root;
                for (int j = i - 1; j >= 0; j--) {
                    char ch = s.charAt(j);
                    if (!node.children.containsKey(ch)) {
                        break;
                    }
                    node = node.children.get(ch);
                    if (dp[j] && node.is_end) {
                        dp[i] = true;
                        break;
                    }
                }
            }
        }

        return dp[s.length()];
    }

    public static void main(String[] args) {
        System.out.println(wordBreakDpTrie("leetcode", Arrays.asList("leet", "code")));  // true
        System.out.println(wordBreakDpTrie("applepenapple", Arrays.asList("apple", "pen")));  // true
        System.out.println(wordBreakDpTrie("catsandog", Arrays.asList("cat", "cats", "and", "sand", "dog")));  // false
    }
}

File: solutions/Rust/word_break_dp_trie.rs
use std::collections::HashMap;

#[derive(Clone)]
struct TrieNode {
    children: HashMap<char, Box<TrieNode>>,
    is_end: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            is_end: false,
        }
    }
}

/**
 * Determine if string can be segmented using words from dictionary (Trie approach).
 *
 * Time Complexity: O(n*m)
 * Space Complexity: O(n + T)
 */
pub fn word_break_dp_trie(s: String, word_dict: Vec<String>) -> bool {
    let mut root = TrieNode::new();

    for word in word_dict {
        let mut node = &mut root;
        for ch in word.chars() {
            node = &mut node.children.entry(ch).or_insert_with(|| Box::new(TrieNode::new()));
        }
        node.is_end = true;
    }

    let mut dp = vec![false; s.len() + 1];
    dp[0] = true;

    for i in 1..=s.len() {
        if !dp[i] {
            let mut node = &root;
            for j in (0..i).rev() {
                if let Some(next) = node.children.get(&s.chars().nth(j).unwrap()) {
                    node = next;
                    if dp[j] && node.is_end {
                        dp[i] = true;
                        break;
                    }
                } else {
                    break;
                }
            }
        }
    }

    dp[s.len()]
}

fn main() {
    println!("{}", word_break_dp_trie("leetcode".to_string(), vec!["leet".to_string(), "code".to_string()]));  // true
}
